### ISSUE TITLE
group holder type for easier defining and initialization of groups repeatedly.

### ISSUE THREAD

**Issue by pgruenbacher** [skypjack/entt#251]
currently entt::as_group has limited functionality, and it's also impossible to define a entt::group with exclusion type lists since the exclusion types aren't actually part of the group type...

I'm thinking something like this

```
template<typename...>
class group_holder;

template<typename GetT, typename ...Owned>
struct group_holder<GetT, Owned...> {
  GetT get;

  template<typename registry_type>
  constexpr auto make(registry_type& reg) {
    return reg.template group<Owned...>(get);
  }
};

template<typename GetT, typename ExcludeT, typename ...Owned>
struct group_holder<GetT, ExcludeT, Owned...> {
  GetT get;
  ExcludeT exclude;

  template<typename registry_type>
  constexpr auto make(registry_type& reg) const {
    return reg.template group<Owned...>(get, exclude);
  }
};

constexpr auto HorseAnimGroup = group_holder<entt::get_t<const Movement::WriteEntityState, Data::UpperBodyAnim>, HorseAnim>();
```

**Comment by pgruenbacher** [comment]
preferably the api would be more like: 

`constexpr auto HorseAnimGroup = group_holder<HorseAnim>(entt::get<const Movement::WriteEntityState, Data::UpperBodyAnim>);`

**Comment by skypjack** [comment]
To be honest, I think I don't get exactly what's the purpose of this tool.  
Just a class that brings in its signature the information about owned, non-owned and excluded types?

**Comment by pgruenbacher** [comment]
a) if user wanted to define a class with a group member type, they can't really because you can't define a group with excluding types, the excluding types are only used by the registry.group() method.
b) it should ideally fix issue with the as_group deduction not working.
c) it should allow users to define a group more easily, and then use that placeholder type to generate the actual group in different functions.

**Comment by skypjack** [comment]
Wouldn't it be enough to _extend_ the definition of a group so as to include also excluded components?  
Something like this:
```
entt::group<exclude_t<T...>, get_t<U...>, V...>
```
Instead of:
```
entt::group<get_t<U...>, V...>
```
It's true that the `T...` parameter pack isn't used internally, but this way the class would bring all the information that are strictly related to the group itself.

**Comment by pgruenbacher** [comment]
that could work too, i assumed there was a reason why it had to be the way it was currently.

**Comment by skypjack** [comment]
Indeed no. Groups don't contain excluded components because the latter aren't necessary to make a group work properly.  
I've never realized that it could have been considered a design flaw because the class itself isn't explicit about how it's constructed.

**Comment by pgruenbacher** [comment]
well the type deduction of the as_group method would still be an issue I suppose.

**Comment by skypjack** [comment]
Fortunately no.  
I made an attempt a while ago and faced an issue of VS. Then I gave up, you're right. I gave without even trying something different.  
However, something different that is also accepted by VS is possible. Look at the last commit on `experimental`. `as_group` now accepts also the observed components.  
It was easy to do actually. Trivial indeed. Probably I was too tired to realize it the first time I tried it. :-)
